File: HYLImageManager/HYLImageManager.m
//
//  HYLImageManager.m
//The MIT License (MIT)
//
//Copyright (c) 2015 Yilei He
//https://github.com/lionhylra/HYLImageManager
//
//Permission is hereby granted, free of charge, to any person obtaining a copy
//of this software and associated documentation files (the "Software"), to deal
//in the Software without restriction, including without limitation the rights
//to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//copies of the Software, and to permit persons to whom the Software is
//furnished to do so, subject to the following conditions:
//
//The above copyright notice and this permission notice shall be included in all
//copies or substantial portions of the Software.
//
//THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
//SOFTWARE.

#import "HYLImageManager.h"
#import "HYLFileManager.h"
#import "UIImage+Compress.h"

@interface HYLImageManager ()
@property(nonatomic, strong) HYLFileManager *fileManager;
@property(nonatomic, strong) HYLFileManager *thumbnailFileManager;
- (NSString *)nameFromTimestamp;
@end

@implementation HYLImageManager

-(instancetype)init{
    return [self initWithDirectory:kDefaultDirectory pathComponents:@[kDefaultPathComponent] maxSize:CGSizeZero thumbnailMaxSize:CGSizeZero compressionQuality:1.0 ignoreThumbnail:YES];
}

-(instancetype)initWithDirectory:(NSSearchPathDirectory)directory{
    return [self initWithDirectory:directory pathComponents:@[kDefaultPathComponent]];
}

-(instancetype)initWithDirectory:(NSSearchPathDirectory)directory pathComponents:(nullable NSArray *)pathComponents{
    return [self initWithDirectory:directory pathComponents:pathComponents maxSize:CGSizeZero];
}

-(instancetype)initWithDirectory:(NSSearchPathDirectory)directory pathComponents:(NSArray *)pathComponents maxSize:(CGSize)maxSize{
    return [self initWithDirectory:directory pathComponents:pathComponents maxSize:maxSize thumbnailMaxSize:CGSizeZero compressionQuality:1.0 ignoreThumbnail:YES];
}

-(instancetype)initWithDirectory:(NSSearchPathDirectory)directory pathComponents:(NSArray *)pathComponents maxSize:(CGSize)maxSize thumbnailMaxSize:(CGSize)thumbnailMaxSize compressionQuality:(float)quality ignoreThumbnail:(BOOL)flag{
    self = [super init];
    if (self) {
        _fileManager = [[HYLFileManager alloc]initWithDirectory:directory rootPathComponents:pathComponents];
        _thumbnailFileManager = [[HYLFileManager alloc]initWithDirectory:directory rootPathComponents:[pathComponents arrayByAddingObject:@"thumbnail"]];
        _maxSize = maxSize;
        _compressionQuality = quality;
        _thumbnailMaxSize = thumbnailMaxSize;
        _ignoreThumbnail = flag;
    }
    return self;
}


+(instancetype)defaultManager{
    /* Singleton */
    static dispatch_once_t pred;
    static id sharedInstance = nil;
    dispatch_once(&pred, ^{
        sharedInstance = [[[self class] alloc] init];
    });
    return sharedInstance;
}

-(NSSearchPathDirectory)directory{
    return self.fileManager.directory;
}

-(NSArray * __nullable)rootPathComponents{
    return self.fileManager.rootPathComponents;
}

-(NSString * __nonnull)pathForImageName:(NSString * __nonnull)imageName{
    return [self pathForImageName:imageName isThumbnail:NO];
}

-(NSString * __nonnull)pathForImageName:(NSString * __nonnull)imageName isThumbnail:(BOOL)isThumbnail{
    if (isThumbnail&&!self.ignoreThumbnail) {
        return [self.thumbnailFileManager pathForFileName:imageName];
    }else{
        return [self.fileManager pathForFileName:imageName];
    }
}

-(BOOL)imageExistsForImageName:(NSString * __nonnull)imageName{
    return [self.fileManager fileExistsForFileName:imageName];
}

-(nullable UIImage *)imageWithName:(NSString * __nonnull)imageName{
    return [self imageWithName:imageName isThumbnail:NO];
}

-(nullable UIImage *)imageWithName:(NSString * __nonnull)imageName isThumbnail:(BOOL)isThumbnail{
    return [UIImage imageWithContentsOfFile:[self pathForImageName:imageName isThumbnail:isThumbnail]];
}

-(NSString * __nonnull)saveImage:(UIImage * __nonnull)image{
    NSString *name = [self nameFromTimestamp];
    [self saveImage:image withImageName:name];
    return name;
}

-(void)saveImage:(UIImage * __nonnull)image withImageName:(NSString * __nonnull)imageName{
    /* resize and compress image */
    UIImage *resizedImage = [UIImage compressedImage:image withMaxWidth:self.maxSize.width maxHeight:self.maxSize.height quality:self.compressionQuality];
    /* write to file */
    NSData *data = UIImageJPEGRepresentation(resizedImage, 1.0);
    [self.fileManager saveData:data withName:imageName];
    if (self.ignoreThumbnail) {
        return;
    }
    /* handle thumbnail */
    UIImage *thumbnailImage = [UIImage compressedImage:image withMaxWidth:self.thumbnailMaxSize.width maxHeight:self.thumbnailMaxSize.width quality:self.compressionQuality*0.9];
    NSData *thumbnailData = UIImageJPEGRepresentation(thumbnailImage, 1.0);
    [self.thumbnailFileManager saveData:thumbnailData withName:imageName];
}

-(BOOL)deleteImageWithImageName:(NSString * __nonnull)imageName error:(NSError *__autoreleasing  __nullable * __nullable)error{
    
    BOOL deleteOriginalSuccess = [self.fileManager deleteFileWithName:imageName error:error];
    if (self.ignoreThumbnail) {
        return deleteOriginalSuccess;
    }
    /* thumbnail */
    BOOL deleteThumbnailSuccess = [self.thumbnailFileManager deleteFileWithName:imageName error:error];
    return deleteOriginalSuccess&&deleteThumbnailSuccess;
}

-(BOOL)renameImageFromImageName:(NSString * __nonnull)oldImageName toNewImageName:(NSString * __nonnull)newImageName error:(NSError *__autoreleasing  __nullable * __nullable)error{
    BOOL renameOriginalSuccess = [self.fileManager renameFileFromFileName:oldImageName toNewFileName:newImageName error:error];
    if (self.ignoreThumbnail) {
        return renameOriginalSuccess;
    }
    /* thumbnail */
    BOOL renameThumbnailSuccess= [self.thumbnailFileManager renameFileFromFileName:oldImageName toNewFileName:newImageName error:error];
    return renameOriginalSuccess && renameThumbnailSuccess;
}

#pragma mark - private methods

- (NSString *)nameFromTimestamp {
    // Get time stamp as file name
    NSString *timestamp = [NSString stringWithFormat:@"%f", [[NSDate date] timeIntervalSince1970] * 1000000];
    
    // Get the integer part to use as file name
    NSString *fileName = [[timestamp componentsSeparatedByString:@"."] firstObject];
    fileName = [fileName stringByAppendingString:@".jpg"];
    return fileName;
}



@end

File: swift/HYLImageManager/HYLImageManager.swift
//
//  HYLImageManager.swift
//  HYLMultiImagesViewControllerDemo
//
//  Created by HeYilei on 21/08/2015.
//  Copyright (c) 2015 HeYilei. All rights reserved.
//

import UIKit

public class HYLImageManager: NSObject {
    // MARK: - Public Properties
    public let maxSize:CGSize
    public let thumbnailMaxSize:CGSize
    public let compressionQuality:CGFloat
    public var rootPathComponents:[String]?{
        return self.fileManager.rootPathComponents
    }
    public var directory:NSSearchPathDirectory{
        return self.fileManager.directory
    }
    public var ignoreThumbnail:Bool = true
    public static let defaultManager = HYLImageManager()
    // MARK: - Private Properties
    private let fileManager:HYLFileManager
    private let thumbnailFileManager:HYLFileManager
    
    // MARK: - Initializers
    convenience override init() {
        self.init(directory: kDefaultDirectory)
    }
    
    convenience init(directory:NSSearchPathDirectory){
        self.init(directory:directory, pathComponents: [kDefaultPathComponent])
    }
    
    convenience init(directory:NSSearchPathDirectory, pathComponents:[String]?){
        self.init(directory:directory, pathComponents: pathComponents, maxSize: CGSizeZero)
    }
    
    convenience init(directory: NSSearchPathDirectory, pathComponents:[String]?, maxSize:CGSize) {
        self.init(directory: directory, pathComponents: pathComponents, maxSize: maxSize, thumbnailMaxSize: CGSizeZero, compressionQuality:1.0, ignoreThumbnail:true)
    }
    
    init(directory: NSSearchPathDirectory, pathComponents: [String]?, maxSize: CGSize, thumbnailMaxSize: CGSize, compressionQuality quality: CGFloat, ignoreThumbnail flag: Bool) {
        self.fileManager = HYLFileManager(directory: directory, rootPathComponents: pathComponents)
        var thumbnailPathComponents = pathComponents
        thumbnailPathComponents?.append("thumbnail")
        self.thumbnailFileManager = HYLFileManager(directory: directory, rootPathComponents:thumbnailPathComponents)
        self.maxSize = maxSize
        self.thumbnailMaxSize = thumbnailMaxSize
        self.compressionQuality = quality
        self.ignoreThumbnail = flag
        super.init()
    }
    
    // MARK: - Image Operation Methods
    public func pathForImageName(imageName:String) -> String? {
        return self.pathForImageName(imageName, isThumbnail: false)
    }
    
    public func pathForImageName(imageName:String, isThumbnail:Bool) -> String?{
        if !self.ignoreThumbnail && isThumbnail {
            return self.thumbnailFileManager.pathForFileName(imageName)
        }else{
            return self.fileManager.pathForFileName(imageName)
        }
    }
    
    public func imageExistsForImageName(imageName:String) -> Bool{
        return self.fileManager.fileExistsForFileName(imageName)
    }
    
    public func imageWithName(imageName:String) -> UIImage? {
        return self.imageWithName(imageName, isThumbnail: false)
    }
    
    public func imageWithName(imageName:String, isThumbnail:Bool) -> UIImage?{
        if let path = self.pathForImageName(imageName, isThumbnail: isThumbnail) {
            return UIImage(contentsOfFile: path)
        }
        return nil
    }
    
    public func saveImage(image:UIImage) -> String{
        let name = self.nameFromTimestamp()
        self.saveImage(image, withImageName: name)
        return name
    }
    
    public func saveImage(image:UIImage, withImageName imageName:String){
        /* resize and compress image */
        let resizedImage = UIImage.compressImage(image, withMaxWidth: self.maxSize.width, maxHeight: self.maxSize.height, quality: self.compressionQuality)
        /* write to file */
        let data = UIImageJPEGRepresentation(image, 1.0)
        self.fileManager.saveData(data, withName: imageName)
        if self.ignoreThumbnail {
            return
        }
        
        /* thumbnail */
        let thumbnailImage = UIImage.compressImage(image, withMaxWidth: self.thumbnailMaxSize.width, maxHeight: self.thumbnailMaxSize.height, quality: self.compressionQuality * CGFloat(0.9))
        let thumbnailData = UIImageJPEGRepresentation(thumbnailImage, 1.0)
        self.thumbnailFileManager.saveData(thumbnailData, withName: imageName)
    }
    
    public func deleteImageWithImageName(imageName:String) throws{
        var error: NSError! = NSError(domain: "Migrator", code: 0, userInfo: nil)
        let didDeleteOriginal: Bool
        do {
            try self.fileManager.deleteFileWithName(imageName)
            didDeleteOriginal = true
        } catch var error1 as NSError {
            error = error1
            didDeleteOriginal = false
        }
        if self.ignoreThumbnail {
            if didDeleteOriginal {
                return
            }
            throw error
        }
        let didDeleteThumbnail: Bool
        do {
            try self.thumbnailFileManager.deleteFileWithName(imageName)
            didDeleteThumbnail = true
        } catch var error1 as NSError {
            error = error1
            didDeleteThumbnail = false
        }
        if didDeleteOriginal && didDeleteThumbnail {
            return
        }
        throw error
    }
    
    public func renameImageFromImageName(imageName:String, toNewImageName newName:String) throws {
        var error: NSError! = NSError(domain: "Migrator", code: 0, userInfo: nil)
        let didRenameOriginal: Bool
        do {
            try self.fileManager.renameFileFrom(imageName, toNewFileName: newName)
            didRenameOriginal = true
        } catch var error1 as NSError {
            error = error1
            didRenameOriginal = false
        }
        if self.ignoreThumbnail {
            if didRenameOriginal {
                return
            }
            throw error
        }
        
        let didRenameThumbnail: Bool
        do {
            try self.thumbnailFileManager.renameFileFrom(imageName, toNewFileName: newName)
            didRenameThumbnail = true
        } catch var error1 as NSError {
            error = error1
            didRenameThumbnail = false
        }
        if didRenameOriginal && didRenameThumbnail {
            return
        }
        throw error
    }
    
    private func nameFromTimestamp() -> String{
        let timestamp = NSString(format: "%f", NSDate().timeIntervalSince1970 * 1000000)
        
        var fileName = timestamp.componentsSeparatedByString(".")[0] 
        fileName = fileName.stringByAppendingPathExtension("jpg")!
        return fileName
    }
}
